client/service/src/lib.rs

```diff
@@ -196,178 +107,80 @@ pub trait AbstractService: Future<Output = Result<(), Error>> + Send + Unpin + S
-		self.transaction_pool.clone()
-	}
-
-	fn on_exit(&self) -> exit_future::Exit {
-		self.task_manager.on_exit()
-	}
-
-	fn prometheus_registry(&self) -> Option<prometheus_endpoint::Registry> {
-		self.prometheus_registry.clone()
-	}
-
-	fn base_path(&self) -> Option<Arc<BasePath>> {
-		self._base_path.clone()
-	}
 }
 
-impl<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> Future for
-	Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc>
-{
-	type Output = Result<(), Error>;
-
-	fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-		let this = Pin::into_inner(self);
+/// Sinks to propagate telemetry connection established events.
+pub struct TelemetryOnConnectSinks(pub Arc<Mutex<Vec<TracingUnboundedSender<()>>>>);
 
-		match Pin::new(&mut this.essential_failed_rx).poll_next(cx) {
-			Poll::Pending => {},
-			Poll::Ready(_) => {
-				// Ready(None) should not be possible since we hold a live
-				// sender.
-				return Poll::Ready(Err(Error::Other("Essential task failed.".into())));
-			}
-		}
-
-		// The service future never ends.
-		Poll::Pending
+impl TelemetryOnConnectSinks {
+	/// Get event stream for telemetry connection established events.
+	pub fn on_connect_stream(&self) -> TracingUnboundedReceiver<()> {
+		let (sink, stream) =tracing_unbounded("mpsc_telemetry_on_connect");
+		self.0.lock().push(sink);
+		stream
 	}
 }
 
-impl<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> Spawn for
-	Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc>
-{
-	fn spawn_obj(
-		&self,
-		future: FutureObj<'static, ()>
-	) -> Result<(), SpawnError> {
-		self.task_manager.spawn_handle().spawn("unnamed", future);
-		Ok(())
-	}
+/// The individual components of the chain, built by the service builder. You are encouraged to
+/// deconstruct this into its fields.
+pub struct ServiceComponents<
+	TBl: BlockT, TBackend: sc_client_api::backend::Backend<TBl>, TExec, TRtApi, TSc, TExPool,
+> {
+	/// A blockchain client.
+	pub client: Arc<Client<TBackend, TExec, TBl, TRtApi>>,
+	/// A shared transaction pool instance.
+	pub transaction_pool: Arc<TExPool>,
+	/// The chain task manager. 
+	pub task_manager: TaskManager,
+	/// A keystore that stores keys.
+	pub keystore: sc_keystore::KeyStorePtr,
+	/// A shared network instance.
+	pub network: Arc<sc_network::NetworkService<TBl, <TBl as BlockT>::Hash>>,
+	/// RPC handlers that can perform RPC queries.
+	pub rpc_handlers: Arc<RpcHandlers>,
+	/// A shared instance of the chain selection algorithm.
+	pub select_chain: Option<TSc>,
+	/// Sinks to propagate network status updates.
+	pub network_status_sinks: NetworkStatusSinks<TBl>,
+	/// A prometheus metrics registry, (if enabled).
+	pub prometheus_registry: Option<prometheus_endpoint::Registry>,
+	/// Shared Telemetry connection sinks,
+	pub telemetry_on_connect_sinks: TelemetryOnConnectSinks,
+	/// A shared offchain workers instance.
+	pub offchain_workers: Option<Arc<sc_offchain::OffchainWorkers<
+		Client<TBackend, TExec, TBl, TRtApi>,
```

Comment:
hopefully, we can use the generic `Client` param here too, instead of the concrete struct.